indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/PendingTaskBasedResourceManagementStrategy.java

```diff
@@ -0,0 +1,360 @@
+        if (workerNodeIds.contains(s)) {
+          stillExisting.add(s);
+        }
+      }
+      currentlyTerminating.clear();
+      currentlyTerminating.addAll(stillExisting);
+
+      if (currentlyTerminating.isEmpty()) {
+        final int scaleDownTarget = getScaleDownTarget(zkWorkers, workerConfig);
+        final Predicate<ZkWorker> isLazyWorker = ResourceManagementUtil.createLazyWorkerPredicate(config);
+        final List<String> laziestWorkerIps =
+            Lists.transform(
+                runner.markWorkersLazy(isLazyWorker, scaleDownTarget),
+                new Function<ZkWorker, String>()
+                {
+                  @Override
+                  public String apply(ZkWorker zkWorker)
+                  {
+                    return zkWorker.getWorker().getIp();
+                  }
+                }
+            );
+        if (laziestWorkerIps.isEmpty()) {
+          log.debug("Found no lazy workers");
+        } else {
+          log.info(
+              "Terminating %,d lazy workers: %s",
+              laziestWorkerIps.size(),
+              Joiner.on(", ").join(laziestWorkerIps)
+          );
+
+          final AutoScalingData terminated = workerConfig.getAutoScaler().terminate(laziestWorkerIps);
+          if (terminated != null) {
+            currentlyTerminating.addAll(terminated.getNodeIds());
+            lastTerminateTime = new DateTime();
+            scalingStats.addTerminateEvent(terminated);
+            didTerminate = true;
+          }
+        }
+      } else {
+        Duration durSinceLastTerminate = new Duration(lastTerminateTime, new DateTime());
+
+        log.info("%s terminating. Current wait time: %s", currentlyTerminating, durSinceLastTerminate);
+
+        if (durSinceLastTerminate.isLongerThan(config.getMaxScalingDuration().toStandardDuration())) {
+          log.makeAlert("Worker node termination taking too long!")
+             .addData("millisSinceLastTerminate", durSinceLastTerminate.getMillis())
+             .addData("terminatingCount", currentlyTerminating.size())
+             .emit();
+
+          currentlyTerminating.clear();
+        }
+      }
+
+      return didTerminate;
+    }
+  }
+
+  private int getScaleDownTarget(Collection<ZkWorker> zkWorkers, WorkerBehaviorConfig workerConfig)
+  {
+    final Predicate<ZkWorker> isValidWorker = ResourceManagementUtil.createValidWorkerPredicate(config);
+    final int currValidWorkers = Collections2.filter(zkWorkers, isValidWorker).size();
+    final int invalidWorkers = zkWorkers.size() - currValidWorkers;
+    final int minWorkers = workerConfig.getAutoScaler().getMinNumWorkers();
+
+    // Scale Down target
+    // All invalid workers + any lazy workers above minCapacity
+    return invalidWorkers + Math.max(
+        0,
+        Math.min(
+            config.getMaxScalingStep(),
+            currValidWorkers - minWorkers
+        )
+    );
+  }
+
+  @Override
+  public ScalingStats getStats()
+  {
+    return scalingStats;
```

Comment:
At this point I don't know what I was looking at, possibly the wrong class. ScalingStats is lock protected so it should be ok however the caller is using it.